doc/RPC_Layer_Core.rst

```diff
@@ -0,0 +1,266 @@
+
+  - Updatestream which contains given RPC-item created / deleted;
+
+  - Given RPC-item processing started / finished;
+
+  - etc.
+  
+Connectivity subcomponent has to provide asynchronous and synchronous functions to send/receive FOPs and bulks, and establish/close connections between end-points via networking layer and provide high level logic abstractions over connections by introducing sessions and slots. Session is a dynamically created server object, created and destroyed by client requests. It maintains the server's state relative to the connection(s) belonging to a client instance. Slot is created by a client request and attached to active client session. Slot is a context in which some operations on RPC items are being applied sequentially one by one (execution for FOPs, write to ADDB for ADDB-records). RPC items related to different slots can be processed concurrently.
+
+Connectivity exposes interfaces:
+
+- Establish/close connection to end-point and return session.
+
+- Create/destroy slot.
+
+- Asynchronous and synchronous RPC-item transmitting functions.
+
+- Specify properties of subcomponent like max slots existing simultaneously.
+
+Further description of connectivity is out of the scope of this document and described in sessions HLD [5].
+
+Statistics subcomponent is used to gather and share various statistics of RPC layer. This can be different vital and miscellaneous information regarding RPC component state and other properties. The following properties look to be important for RPC core component user:
+
+- Items in caches and their count;
+
+- Items in formation stage;
+
+- RPCs count in output stage;
+
+- Count of processing threads and their state;
+
+- Number of opened sessions and slots;
+
+- Time related statistics:
+
+  - Min/max/average time of RPC-item processing;
+
+  - Utilization of various resources (% of max network bandwidth used);
+
+  - Usage ranking statistics (79% of RPCs were high-priority, 4% RPCs came from client X, etc.);
+
+  - RPCs per second;
+
+  - Some units measured in bytes/sec or pages/sec;
+
+- Some constant and slow varying properties like C2_MAX_PAGES_PER_RPC or C2_MAX_RPCS_IN_FLIGHT.
+
+Statistics subcomponent expose simple interface which takes property name or key and returns its value in raw form.
+
+Interface description:
+
+Component IF:
+
+- int c2_rpc_core_init(...);
+
+- void c2_rpc_core_fini();
+
+Statistics IF:
+
+- void* c2_rpc_stat_get_prop();
+
+Connectivity IF:
+
+- c2_rpc_session_id c2_rpc_session_link(c2_service_id);
+
+- c2_rpc_slot_id c2_rpc_slot_link(c2_rpc_session_id);
+
+- void c2_rpc_session_unlink(c2_rpc_session_id);
+
+- void c2_rpc_slot_unlink(c2_rpc_slot_id);
+
+- void c2_rpc_slot_max_set(uint32_t max);
+
+Processing IF:
+
+- int c2_rpc_item_send(c2_rpc_item*, c2_service_id, c2_rpc_prio, c2_rpc_caching_type, c2_rpc_callback*);
```

Comment:
rename all "c2_" prefix to "m0_" prefix.

Reply:
Incorporated.